cat/src/Main.gren

```diff
@@ -19,36 +19,50 @@ main =
 
 
 type alias Model =
-    { stdout : Stream }
+    { stdout : Stream
+    }
 
 
 type Msg
     = OpenResult (Result FileSystem.AccessError (FileSystem.ReadableFileHandle Never))
     | ReadResult (Result FileSystem.UnknownFileSystemError Bytes)
 
 
-init : Program.AppInitTask { model : Model, command : Cmd Msg }
+init :
+    Program.AppInitTask
+        { model : Model
+        , command : Cmd Msg
+        }
 init =
-    Program.await Node.initialize <| \nodeConfig ->
-        Program.await FileSystem.initialize <| \fsPermission ->
-            Program.startProgram
-                { model =
-                    { stdout = nodeConfig.stdout
-                    }
-                , command =
-                    case nodeConfig.args of
-                        [ _, _, file ] ->
-                            FileSystem.openForRead fsPermission file
-                                |> Task.attempt OpenResult
-
-                        _ ->
-                            Stream.send nodeConfig.stderr <|
-                                BE.encode <|
-                                    BE.string "Exactly one argument is required: the file name to read\n"
-                }
-
-
-update : Msg -> Model -> { model : Model, command : Cmd Msg }
+    Program.await Node.initialize
+        <| (\nodeConfig ->
+                Program.await FileSystem.initialize
+                    <| (\fsPermission ->
+                            Program.startProgram
+                                { model =
+                                    { stdout = nodeConfig.stdout
+                                    }
+                                , command =
+                                    case nodeConfig.args of
+                                        [ _, _, file ] ->
+                                            FileSystem.openForRead fsPermission file
+                                                |> Task.attempt OpenResult
+
+                                        _ ->
+                                            Stream.send nodeConfig.stderr
+                                                <| BE.encode
+                                                <| BE.string "Exactly one argument is required: the file name to read\n"
+                                }
+                        )
+            )
```

Comment:
I hope we can format this block of code in a nicer way.

1. `<|` seems to always be split out to a seperate line, I'd usually prefer it to be on the same line. I believe that's what happens in `elm-format` too, or am I wrong?
2. As with `<|`, I prefer lambdas to stay on the previous line, instead of being broken out to a separate line. Not sure what `elm-format` does in this case.
3. Wrapping a lambda in parenthesis should be uneccessary when on the right side of `<|`.

While I'm giving this feedback to this particular block of code, my comments apply universally.

Personally, I'd love to see the `await` statements formatted to be on the same line, but I appriciate there being different views on this and it's not important to get in before the 0.2 release.

Reply:
:+1: Sounds good.  I already was expecting this from following some of the discussions on Zulip :D  I think this will be doable, but as it's quite different from elm-format, it'll take a bit of thinking through the details of the rules, so I've postponed worrying about it yet.

(1): Yes, elm-format will keep it on a single line if it's not otherwise forced to expand.  For gren format, retaining information about where linebreaks originally were isn't implemented at all yet, so allowing single-line `<|` would fit as part of that future goal.  I guess the main question here is, is `<|` going to be considered special, or do we want a rule that applies to all binary operators?  (For elm-format, `<|` is in fact a special case, see below.)

(2): elm-format doesn't allow that, though it does have one special case for `<| \_ -> ...` where the `<|` will stick at the end of the previous line (but the lambda will still start on the next line).  This sounds fine to try to implement.

(3): I think I made it add parens in the first pass because I wasn't confident that I understood the edge cases well enough to be confident that it wouldn't remove parens in some cases that are actually necessary.  (Since the parser originally dropped all parens when parsing, the formatter actually is only adding parens that are required -- which in total has the result of removing unnecessary parens.)  This should be fine to implement.  Though if anyone happens to have some time to help enumerate all the possible edge cases where parens may or may not be required around lambdas to avoid changing the meaning of code, that'd be helpful!